File: algorithms/2_medium/R/00788_rotated_digits/safe_static.c
#define MAX_ALLOWED_N 10000
#define VALIDS_LEN 3
#define VALIDS_WITH_DIFF_LEN 4

static const int8_t VALIDS[VALIDS_LEN] = { 0, 1, 8 };
static const int8_t VALIDS_WITH_DIFF[VALIDS_WITH_DIFF_LEN] = { 2, 5, 6, 9 };

static pthread_once_t once_init = PTHREAD_ONCE_INIT;
static int16_t freqs[MAX_ALLOWED_N + 1] = { 0 };

void init_backtrack(const int a, const bool contains_diff)
{
    if (a > MAX_ALLOWED_N) {
        return;
    }

    if (contains_diff) {
        freqs[a] = 1;
    }

    const int base = a * 10;

    for (size_t i = 0; i < VALIDS_LEN; i += 1) {
        init_backtrack(base + VALIDS[i], contains_diff);
    }

    for (size_t i = 0; i < VALIDS_WITH_DIFF_LEN; i += 1) {
        init_backtrack(base + VALIDS_WITH_DIFF[i], true);
    }
}

void init(void)
{
    for (size_t i = 0; i < VALIDS_LEN; i += 1) {
        if (VALIDS[i] != 0) {
            init_backtrack(VALIDS[i], false);
        }
    }

    for (size_t i = 0; i < VALIDS_WITH_DIFF_LEN; i += 1) {
        init_backtrack(VALIDS_WITH_DIFF[i], true);
    }

    for (size_t i = 1; i <= MAX_ALLOWED_N; i += 1) {
        freqs[i] += freqs[i - 1];
    }
}

int rotatedDigits(int n)
{
    assert(n <= MAX_ALLOWED_N && n >= 1);
    pthread_once(&once_init, init);
    return freqs[n];
}

File: algorithms/2_medium/R/00788_rotated_digits/safe_static.rs
use std::sync::Once;

static INIT: Once = Once::new();
const MAX_ALLOWED_N: i32 = 10_000;
const VALIDS: [i8; 3] = [0, 1, 8];
const VALIDS_WITH_DIFF: [i8; 4] =  [ 2, 5, 6, 9 ];
static mut FREQS: [i16; MAX_ALLOWED_N as usize + 1] = [0; MAX_ALLOWED_N as usize + 1];

fn init_backtrack(a: i32, contains_diff: bool) {
    if a > MAX_ALLOWED_N {
        return;
    }

    if contains_diff {
        unsafe { FREQS[a as usize] = 1; }
    }

    let base = a * 10;

    for v in VALIDS {
        init_backtrack(base + v as i32, contains_diff);
    }

    for v in VALIDS_WITH_DIFF {
        init_backtrack(base + v as i32, true);
    }
}

fn init() {
    for v in VALIDS {
        if (v != 0) {
            init_backtrack(v as i32, false);
        }
    }

    for v in VALIDS_WITH_DIFF {
        init_backtrack(v as i32, true);
    }

    for i in 1..=(MAX_ALLOWED_N as usize) {
        unsafe { FREQS[i] += FREQS[i - 1]; }
    }
}


impl Solution {
    pub fn rotated_digits(mut n: i32) -> i32 {
        assert!(n <= MAX_ALLOWED_N && n >= 1);

        INIT.call_once(|| {
            init();
        });

        unsafe { FREQS[n as usize] as i32 }
    }
}
